File: Ceballos-Conde-Emanuel/Project euler/semiprimos.py
import math

limite = 10**8
p_limite = limite//2
primos = []
es_primo = [True]*(p_limite+1)

for p in range(2, p_limite + 1):
    if es_primo[p]:
        primos.append(p)
        for i in range(p * p, p_limite + 1, p):
            es_primo[i] = False

numero_de_primos= len(primos)
num_semiprimos = 0
puntero_derecho = numero_de_primos -1

for puntero_izquierdo in range(numero_de_primos):
    p = primos[puntero_izquierdo]

    if (p*p) > limite:
        break
    
    while puntero_derecho >= puntero_izquierdo and p*primos[puntero_derecho]>=limite:
        puntero_derecho -=1
    
    if puntero_derecho >= puntero_izquierdo:
        num_semiprimos += (puntero_derecho - puntero_izquierdo +1)

print(num_semiprimos)

File: Ceballos-Conde-Emanuel/Project euler/semiprimos.c
#include <stdio.h>
#include <stdlib.h>

int main() {
    long long limite = 100000000LL;
    long long p_limite = limite / 2;

    unsigned char *no_es_primo = (unsigned char *)calloc(p_limite + 1, sizeof(unsigned char));
    
    for (long long p = 2; p * p <= p_limite; p++) {
        if (!no_es_primo[p]) {
            for (long long i = p * p; i <= p_limite; i += p)
                no_es_primo[i] = 1;
        }
    }

    long long *primos = (long long *)malloc((p_limite / 2) * sizeof(long long)); 
    long long num_primos = 0;
    for (long long p = 2; p <= p_limite; p++) {
        if (!no_es_primo[p]) {
            primos[num_primos++] = p;
        }
    }
    free(no_es_primo);

    long long num_semiprimos = 0;
    long long der = num_primos - 1;

    for (long long izq = 0; izq < num_primos; izq++) {
        long long p = primos[izq];
        if (p * p >= limite) break;
        while (der >= izq && p * primos[der] >= limite) {
            der--;
        }
        if (der >= izq) {
            num_semiprimos += (der - izq + 1);
        }
    }

    printf("%lld\n", num_semiprimos);

    free(primos);
    return 0;
}